### PR TITLE
COMPONENTS: New Lab Dialog

### PR DESCRIPTION
**Pull request by terrypalo**
Happy:
image

Loading:
image

Validation Error:
image

Dependency Error:
image

### CODE REVIEW

File: DMX.Portal.Web.Tests.Unit/Components/NewLabDialogs/NewLabDialogComponentTests.Render.cs [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            // when 
+            this.renderedNewLabDialog = RenderComponent<NewLabDialog>();
+            this.renderedNewLabDialog.Instance.OpenDialog();
+
+            // then
+            this.renderedNewLabDialog.Instance.State.Should().Be(expectedState);
+            this.renderedNewLabDialog.Instance.LabViewService.Should().NotBeNull();
+            this.renderedNewLabDialog.Instance.Dialog.Should().NotBeNull();
+            this.renderedNewLabDialog.Instance.Dialog.IsVisible.Should().BeTrue();
+            this.renderedNewLabDialog.Instance.Dialog.Title.Should().Be("Lab Details");
```
Comment (hassanhabib) [inline DMX.Portal.Web.Tests.Unit/Components/NewLabDialogs/NewLabDialogComponentTests.Render.cs:71]: Create a variable and call it `expectedDialogTitle` - do the same for all the other hard coded values in your test.

File: DMX.Portal.Web/Views/Bases/TextBoxBase.razor.cs [hunk 2]
```diff
@@ -21,13 +21,20 @@ public partial class TextBoxBase : ComponentBase
...
-        {
-            this.Value = value;
-            await ValueChanged.InvokeAsync(this.Value);
-        }
+        public Task SetValueAsync(string value) =>
+            InvokeAsync(async () =>
+            {
+                this.Value = value;
+                await this.ValueChanged.InvokeAsync(this.Value);
+            });
```
Comment (hassanhabib) [inline DMX.Portal.Web/Views/Bases/TextBoxBase.razor.cs:37]: Pull this guy back by one - the same way we do `TryCatch`

File: DMX.Portal.Web/Views/Components/NewLabDialogs/NewLabDialog.razor [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    <Content>
+        <div>
+            <ContainerBase @ref=LabIdContainer
+                           CssClass="new-lab-textbox">
+
+                <TextBoxBase @ref=LabId
+                             @bind-Value="@labIdString"
+                             Placeholder=@nameof(LabView.Id)
+                             IsDisabled=true />
+
```
Comment (hassanhabib) [inline DMX.Portal.Web/Views/Components/NewLabDialogs/NewLabDialog.razor:19]: Too many empty lines - just one

File: DMX.Portal.Web.Tests.Unit/Components/NewLabDialogs/NewLabDialogComponentTests.Exceptions.cs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+            this.renderedNewLabDialog.Instance.LabNameValidationSummary.ValidationData
+                .Should().BeEquivalentTo(
+                    labViewValidationException.InnerException.Data);
+
+            this.renderedNewLabDialog.Instance.LabNameValidationSummary.Key
+                .Should().Be(nameof(this.renderedNewLabDialog.Instance.LabView.Name));
+
+            this.renderedNewLabDialog.Instance.LabNameValidationSummary.Color
+                .Should().Be("Red");
```
Comment (hassanhabib) [inline DMX.Portal.Web.Tests.Unit/Components/NewLabDialogs/NewLabDialogComponentTests.Exceptions.cs:70]: Put all these hardcoded values into `expectedSomething` variable

File: DMX.Portal.Web.Tests.Unit/Components/NewLabDialogs/NewLabDialogComponentTests.Exceptions.cs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            await this.renderedNewLabDialog.Instance.LabName
+                .SetValueAsync(someLabName);
+
+            await this.renderedNewLabDialog.Instance.LabDescription
+                .SetValueAsync(someLabDescription);
+
+            this.renderedNewLabDialog.Instance.Dialog
+                .Click();
+
+
```
Comment (hassanhabib) [inline DMX.Portal.Web.Tests.Unit/Components/NewLabDialogs/NewLabDialogComponentTests.Exceptions.cs:119]: Remove extra empty line

File: DMX.Portal.Web.Tests.Unit/Components/NewLabDialogs/NewLabDialogComponentTests.Render.cs [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+            this.renderedNewLabDialog.Instance.LabIdContainer.Should().NotBeNull();
+            this.renderedNewLabDialog.Instance.LabIdContainer.CssClass.Should().Be(expectedTextBoxCssClass);
+            this.renderedNewLabDialog.Instance.LabDmxVersionContainer.Should().NotBeNull();
+            this.renderedNewLabDialog.Instance.LabDmxVersionContainer.CssClass.Should().Be(expectedTextBoxCssClass);
+            this.renderedNewLabDialog.Instance.LabNameContainer.Should().NotBeNull();
+            this.renderedNewLabDialog.Instance.LabNameContainer.CssClass.Should().Be(expectedTextBoxCssClass);
+            this.renderedNewLabDialog.Instance.LabDescriptionContainer.Should().NotBeNull();
+            this.renderedNewLabDialog.Instance.LabDescriptionContainer.CssClass.Should().Be(expectedTextBoxBottomCssClass);
+            this.renderedNewLabDialog.Instance.DependencyErrorMessageLabel.Should().NotBeNull();
+            this.renderedNewLabDialog.Instance.DependencyErrorMessageLabel.ClassName.Should().Be(expectedDependencyErrorMessageLabelCssClass);
```
Comment (hassanhabib) [inline DMX.Portal.Web.Tests.Unit/Components/NewLabDialogs/NewLabDialogComponentTests.Render.cs:98]: Keep all lines under 120 characters

File: DMX.Portal.Web/Views/Components/NewLabDialogs/NewLabDialog.razor [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                             Placeholder=@nameof(LabView.Description)
+                             Multiline=true />
+
+                <ValidationSummaryBase @ref=LabDescriptionValidationSummary
+                                       Key=@nameof(LabView.Description)
+                                       ValidationData=@Exception?.Data
+                                       Color="Red" />
+
+            </ContainerBase>
+            <LabelBase @ref="@DependencyErrorMessageLabel"
```
Comment (hassanhabib) [inline DMX.Portal.Web/Views/Components/NewLabDialogs/NewLabDialog.razor:61]: Put a new empty line above this one

File: DMX.Portal.Web/Views/Components/NewLabDialogs/NewLabDialog.razor.Exceptions.cs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            {
+                RenderDependencyError(labViewDependencyException);
+            }
+            catch (LabViewServiceException labViewServiceException)
+            {
+                RenderDependencyError(labViewServiceException);
+            }
+        }
+
+        private void RenderValidationError(Xeption xeption)
```
Comment (hassanhabib) [inline DMX.Portal.Web/Views/Components/NewLabDialogs/NewLabDialog.razor.Exceptions.cs:35]: Call the variable `exception`

**Comment by AndrewMcClelland** [comment]
@AndrewMcClelland & @terrypalo:

- Lab Details -> New Lab
- Save New Lab -> Submit
- Try moving DMXVersion down below name/description
- Run end to end against live API

**Review by hassanhabib — APPROVED** [review]

**Pull request merged**